src/Compilers/VisualBasic/Portable/Binding/Binder_Symbols.vb

```diff
@@ -690,6 +679,73 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                                                    binder As Binder,
+                                                    diagnostics As DiagnosticBag,
+                                                    suppressUseSiteError As Boolean,
+                                                    inGetTypeContext As Boolean,
+                                                    resolvingBaseType As Boolean) As TypeSymbol
+
+                Dim numElements As Integer = syntax.Elements.Count
+                Dim types = ArrayBuilder(Of TypeSymbol).GetInstance(numElements)
+                Dim locations = ArrayBuilder(Of Location).GetInstance(numElements)
+                Dim elementNames As ArrayBuilder(Of String) = Nothing
+
+                ' set of names already used
+                Dim uniqueFieldNames = PooledHashSet(Of String).GetInstance()
+                Dim countOfExplicitNames As Integer = 0
+
+                For i As Integer = 0 To numElements - 1
+                    Dim argumentSyntax = syntax.Elements(i)
+
+                    Dim argumentType As TypeSymbol = binder.BindTypeSyntax(argumentSyntax.Type, diagnostics, suppressUseSiteError, inGetTypeContext, resolvingBaseType)
+                    types.Add(argumentType)
+
+                    If argumentType.IsRestrictedType() Then
+                        Binder.ReportDiagnostic(diagnostics, argumentSyntax, ERRID.ERR_RestrictedType1, argumentType)
+                    End If
+
+                    Dim name As String = Nothing
+                    Dim nameSyntax As IdentifierNameSyntax = argumentSyntax.IdentifierName
+
+                    If nameSyntax IsNot Nothing Then
+                        name = nameSyntax.Identifier.ValueText
+
+                        ' validate name if we have one
+                        countOfExplicitNames += 1
+                        Binder.CheckTupleMemberName(name, i, nameSyntax, diagnostics, uniqueFieldNames)
+                        locations.Add(nameSyntax.GetLocation)
+                    Else
+                        locations.Add(argumentSyntax.GetLocation)
+                    End If
+
+                    Binder.CollectTupleFieldMemberNames(name, i + 1, numElements, elementNames)
+                Next
+
+                uniqueFieldNames.Free()
+
+                If countOfExplicitNames <> 0 AndAlso countOfExplicitNames <> numElements Then
+                    Binder.ReportDiagnostic(diagnostics, syntax, ERRID.ERR_TupleExplicitNamesOnAllMembersOrNone)
+                End If
+
+                Dim typesArray As ImmutableArray(Of TypeSymbol) = types.ToImmutableAndFree()
+                Dim locationsArray As ImmutableArray(Of Location) = locations.ToImmutableAndFree()
+
+                If typesArray.Length < 2 Then
+                    'PROTOTYPE: tuples error ID
+                    Dim diaginfo = diagnostics.Add(ERRID.ERR_TupleTooFewElements, syntax.GetLocation)
+                    Return ErrorTypeSymbol.UnknownResultType
```

Comment:
C# part is fixed in a different enlistment. I will make a separate PR for C# cleanups that came from looking at that code while doing VB

---
In reply to: [71450858](https://github.com/dotnet/roslyn/pull/12596#discussion_r71450858) [](ancestors = 71450858,71443319)